src/main.cpp

```diff
@@ -4502,6 +4534,33 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
     }
 
 
+    // Disconnect existing peer connection if:
+    // 1. The version message has been received
+    // 2. Overwinter is active
+    // 3. Peer version is legacy (pre-Overwinter)
+    // 4. Peer branch id is not supported
+    // FIXME: one branch id or set of branch ids?
+    else if (
+        pfrom->nVersion != 0 &&
+        NetworkUpgradeActive(GetHeight(), chainparams.GetConsensus(), Consensus::UPGRADE_TESTDUMMY))
+    {
+        if (pfrom->nVersion < OVERWINTER_PROTO_VERSION) {
```

Comment:
This should be fine, as we can be sure that any version below this is undesired, regardless of what branch it is from.